ws/api/endpoints.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/mainflux/mainflux/ws"
+)
+
+var (
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			switch err {
+			case errUnauthorizedAccess:
+				w.WriteHeader(http.StatusForbidden)
+				return
+			default:
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err.Error()))
+				w.WriteHeader(http.StatusServiceUnavailable)
+				return
+			}
+		}
+
+		getSubstopic(req, r, w)
+
+		// Upgrade to a new ws connection.
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+
+		// Configure the closing of the ws connection
+		conn.SetCloseHandler(func(code int, text string) error {
+			return svc.Unsubscribe(context.Background(), req.thingKey, req.chanID, req.subtopic)
+		})
+
+		req.conn = conn
+		client := ws.NewClient(conn, "")
+
+		// Subscribe using the adapterservice
+		if err := svc.Subscribe(context.Background(), req.thingKey, req.chanID, req.subtopic, client); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+			if err == ws.ErrFailedConnection {
+				req.conn.Close()
+				return
+			}
+		}
+
+		logger.Debug(fmt.Sprintf("Successfully upgraded communication to WS on channel %s", req.chanID))
+
+		msgs := make(chan messaging.Message)
+
+		// Listen for messages received from the chan messages, and publish them to broker
+		go publish(svc, msgs, req.thingKey)
+
+		go req.broadcast(msgs)
+	}
+}
+
+func publish(svc ws.Service, msgs <-chan messaging.Message, thingKey string) {
+	for msg := range msgs {
+		svc.Publish(context.Background(), thingKey, msg)
+	}
+}
+
+func getSubstopic(req publishReq, r *http.Request, w http.ResponseWriter) {
+	channelParts := channelPartRegExp.FindStringSubmatch(r.RequestURI)
+	if len(channelParts) < 2 {
+		logger.Warn("Empty channel id or malformed url")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	subtopic, err := parseSubTopic(channelParts[2])
+	if err != nil {
+		logger.Warn("Empty channel id or malformed url")
+		w.WriteHeader(http.StatusBadRequest)
+	}
+
+	req.subtopic = subtopic
+}
+
+func parseSubTopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", errMalformedSubtopic
+	}
+
+	subtopic = strings.Replace(subtopic, "/", ".", -1)
+
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if len(elem) > 1 && (strings.Contains(elem, "*") || strings.Contains(elem, ">")) {
+			return "", errMalformedSubtopic
+		}
+
+		filteredElems = append(filteredElems, elem)
+	}
+
+	subtopic = strings.Join(filteredElems, ".")
+
+	return subtopic, nil
+}
```

Comment:
Use a single function to parse subtopic. You can use https://github.com/mainflux/mainflux/blob/6bd745b658ff1e04c40394a732c731fa99882c32/http/api/transport.go#L71 as a refenerce.